ZON-4007 Make payloads configurable via templates

knutgoetz opened the pull request:
Kann wahrscheinlich noch nicht gemergt werden.
Vielleicht nicht der schönste PR, aber zumindest mal eine Diskussionsgrundlage. Einige TODOS und Fragen sind in den Kommentaren.

Commit:
Add mobile payload template field and its source.

Commit:
Add test for new field.

Broke test_applies_push_configuration_to_added_object though.

Commit:
xMerge branch 'master' of ssh://github.com/ZeitOnline/zeit.push into ZON-4007

Commit:
Add test JSON FIle

Commit:
Add a payload source and its tests

Commit:
Delete obsolte testclass

Commit:
Store the token in DAV-property not the whole z.c.text object

Commit:
Add push template basic functionality.

Sorry this commit is quite messy. But i want to deliver it to vivi-devel in order to get some feedback on this first draft.

Commit:
Use urbanairship web credentials

Commit:
Dont use UA webcredentials in tests right now

Commit:
Fix wrong unpacking of push message object and credentials

Commit:
Log template used in object log not the channel

Commit:
Fix setting audiance in push object

Commit:
Move test_push_works test

Commit:
Dont cache in jinja environment

Commit:
Clean up code

Commit:
Change the payload-template so it is valid, using UAs validate endpoint.

Commit:
Make creating a test payload template more flexible.

Commit:
Use UA web credentials in tests

Commit:
Fix or delete all tests.

Commit:
Refactor tests.

Commit:
Delete wrong commentary

Commit:
Adapt selenium tests and add slim payload jinja test.

Commit:
Change CHANGES.xt

Commit:
Add simple docstring to payload-documentation

Commit:
Remove entry point for parse payload documentation

Commit:
Add new push documentation using a template

Commit:
pep8

Commit:
Set 'push_news' flag independent of channel.

Commit:
Rename payload_template field to 'mobile_template_field'

wosc commented on src/zeit/push/__init__.py:
```diff
@@ -17,6 +17,7 @@
   channel-news News
   urbanairship-audience-group subscriptions
   mobile-buttons file://{fixtures}/mobile-buttons.xml
+  push-payload-templates data/payload-templates
```
Hat jetzt mit dem Code nichts zu tun :), aber ich fänds glaub ich cool wenn der Ordner in Production irgendwas mit "push" und/oder "urban airship" heißen würde. Denn `payload` ist ja allgemein "Nutzlast", da wird nicht direkt klar, worums geht. `/data/urbanairship-templates` vielleicht?

knutgoetz replied:
Jo, sehe ich auch so. Änder ich mal fix.

wosc commented on src/zeit/push/browser/tests/test_form.py:
```diff
@@ -10,6 +12,14 @@ class SocialFormTest(zeit.cms.testing.BrowserTestCase):
+            with zeit.cms.testing.interaction():
+                self.repository['data'] = Folder()
+                self.repository['data']['payload-templates'] = Folder()
+                self.repository['data']['payload-templates']['foo.json'] = \
```
Will das nicht evtl. lieber in einen Layer (z.B. [so wie hier](https://github.com/ZeitOnline/zeit.content.dynamicfolder/blob/master/src/zeit/content/dynamicfolder/testing.py#L32))?

wosc commented on src/zeit/push/interfaces.py:
```diff
@@ -244,6 +247,34 @@ class MobileButtonsSource(zeit.cms.content.sources.XMLSource):
+class PayloadTemplateSource(zc.sourcefactory.basic.BasicSourceFactory):
+
+    # For testing purposes, this should not be cached at all
+    # @CONFIG_CACHE.cache_on_arguments()
```
Wie evtl. schonmal gesagt, wenn die Werte der Source ICMSContent-Objekte sind, kommen die ja schon ausm DAV-Cache, d.h. da nochmal in der Source zu cachen ist glaube ich überflüssig.

wosc commented on src/zeit/push/interfaces.py:
```diff
@@ -244,6 +247,34 @@ class MobileButtonsSource(zeit.cms.content.sources.XMLSource):
+            'zeit.push').get('push-payload-templates')
+        template_folder = zeit.cms.interfaces.ICMSContent(
+            "http://xml.zeit.de/{}".format(payload_path))
+        return [template_json for template_json in template_folder.values()]
```
bisschen kürzer und idiomatischer zum Generator auflösen ist `return list(template_folder.values())`

wosc commented on src/zeit/push/interfaces.py:
```diff
@@ -244,6 +247,34 @@ class MobileButtonsSource(zeit.cms.content.sources.XMLSource):
+        payload_path = zope.app.appsetup.product.getProductConfiguration(
+            'zeit.push').get('push-payload-templates')
+        template_folder = zeit.cms.interfaces.ICMSContent(
+            "http://xml.zeit.de/{}".format(payload_path))
```
Wir könnten auch die ganze uniqueId des Ordners in die Konfig tun, nicht nur den Pfad, dann muss man hier nicht rumbasteln.

wosc commented on src/zeit/push/interfaces.py:
```diff
@@ -244,6 +247,34 @@ class MobileButtonsSource(zeit.cms.content.sources.XMLSource):
+    def getToken(self, value):
+        return value.uniqueId
+
+    def find(self, id):
```
Also wenn wir die uniqueId in der message_config speichern, dann implementiert sich das hier flotter als `return ICMSContent(id, None)`. :) Aber wir sollten glaube ich lieber nur den Dateinamen speichern, weil dann der Pfad nicht in jedem Artikel drinsteht. Dann implementiert sich das als `folder.get(id, None)` (oder `ICMSContent(folder_id + id, None)`, wenn wir in der Konfig nen trailing slash verlangen :).

wosc commented on src/zeit/push/message.py:
```diff
@@ -319,3 +319,20 @@ def mobile_buttons(self, value):
         self.push.set(dict(
             type='mobile', channels=zeit.push.interfaces.CONFIG_CHANNEL_NEWS),
```
Der ganze Kram mit `channels=` muss jetzt dann raus, weil der ja vom Template kommt. Das war bisher nur, um Eilmeldung und Wichtige Nachrichten auseinanderzuhalten, weil man ersteres mit "normalen" Artikeln bisher gar nicht pushen konnte.

wosc commented on src/zeit/push/testing.py:
```diff
@@ -27,6 +29,7 @@ def send(self, text, link, **kw):
             text=text, link=link, kw=kw))
 
 
+# TODO Is this still relevant? I guess not.
```
Nope, channel kommt jetzt ja ausm Template. Weg damit. :)

wosc commented on src/zeit/push/urbanairship.py:
```diff
@@ -252,6 +215,8 @@ def text(self):
         result = {
             'mobile_title': self.config.get('title'),
+            'context': self.context,
+            'push_config': self.config
```
Ich frage mich ja, ob wir nicht besser direkt `self` an das send() geben, und der sich dann im create_payload davon runterkratzen kann, was er braucht. Weil so pfriemeln wir zweimal drin rum, einmal hier und dann dort nochmal.

wosc commented on src/zeit/push/urbanairship.py:
```diff
@@ -275,9 +240,11 @@ def product_config(self):
 def set_push_news_flag(context, event):
     push = zeit.push.interfaces.IPushMessages(context)
     for service in push.message_config:
+        # Just ignore channels for now
```
Hmm ich glaub das gibt Ärger. Ich würde eher sagen wir müssen dann statt dem Boolean ein neues solr-indiziertes Feld "push-ua-channel" oder vermutlich noch passender "push-ua-payload" einführen (wie auch immer wir's dann genau nennen).

knutgoetz replied:
Nach Rücksprache mit Thomas kann dass sogar gelöscht werden. Die CP, die mit solchen Artikeln gefüllt wird hat nur den Zweck, dass damit für alte Appversionen ein Link auf die CP gepusht werden kann (aus irgendwelchen Gründen konnten dort keine Links auf die Artikel selbst gepusht werden). Diese alten Versionen sind aber wohl kaum noch in Umlauf und es gibt auch für diese einen Workaround mit der neuen Templatelogik.

wosc commented on src/zeit/push/urbanairship.py:
```diff
@@ -37,149 +38,103 @@ class Connection(object):
-        }
+        article = kw.get('context')
+        push_message = kw.get('push_config')
+        template = self.jinja_env.get_template(push_message.get(
```
Das würde sich glaub ich schon nochmal etwas netter aufschreiben, wenn wir einen eigenen Contenttyp hätten. (Und die ganze Message übergeben würden, siehe unten). Dann könnten wir hier nämlich direkt mit dem Contentobjekt arbeiten, und das könnte direkt eine render-Methode haben. Und das ganze mit dem Environment und dem load_template könnte man im Contenttyp kapseln.

wosc reviewed (commented):
Hier meine Anmerkungen beim ersten Drüberkucken. :)

Commit:
Delete code concerning channels.

Commit:
Add UATemplatelayer to testsetup

Commit:
Rename payload template folder

Commit:
Refactor payload source

Commit:
Dont use strict json parsing for the template.

Commit:
Use message object instead of context and config as additional paramter

Commit:
Delete Code for push_news flag

wosc committed:
Simplify spelling

wosc committed:
Simplify spelling

wosc committed:
ZON-4007: Declare dependency (belongs to commit:6791185)

wosc committed:
Remove dead code

wosc committed:
ZON-4007: Remove tracking link code, this will now be done explicitly in the template

wosc committed:
Shorten log wording (needs to somehow work untranslated in a German message)

wosc committed:
Remove superfluous parameter juggling (belongs to commit:0018fea)

wosc committed:
ZON-4007: Simplify spelling by using jinja's spontaneous environment feature

This caches the environment instance internally, so it's not any more expensive.

wosc committed:
ZON-4007: Move template rendering functionality to the Message class

This enables us to move it further onto a specialized Text content type

wosc committed:
Remove dead code

wosc committed:
Use things we already have (belongs to commit:acbf078)

wosc committed:
ZON-4007: Simplify protocol between Message and send() even further (belongs to commit:91363f5)

wosc committed:
ZON-4007: Add available template variables to documentation

Also, make image_url available in the first place.

wosc committed:
ZON-4007: Add available article properties to documentation

wosc committed:
ZON-4007: Move jinja handling to specialized text content type
